### PR TITLE
Travis fix: Fixes all osx builds and some linux ones

### PR DESCRIPTION
**Pull request by Uditgulati**
Fixes:
- Ruby version (older ones) and bundler version conflicts.
- Errors due to homebrew/science migration to homebrew/core
- yN prompt timeout

Also, updated the corresponding ruby 2.1.x, 2.2.x, 2.3.x, 2.4.x to latest versions and removed non-clang and osx pairs.

There is still some exception (Illegal seek - <STDOUT>) generated in some linux builds while running tests that needs to be fixed. I'll try to fix that too and make a different PR for that.

### CODE REVIEW

File: travis.sh [hunk 1]
```diff
@@ -45,12 +45,18 @@ then
...
     )
 
-    gem pristine --all
-    gem update --no-document --system
-    gem update --no-document
+    yes | gem pristine --all
+    yes | gem update --no-document --system --force
+    yes | gem update --no-document --force
+
+    if [ "$ruby_version" == "2.2.7" ] || [ "$ruby_version" == "2.1.10" ]; then
```
Comment (Shekharrajak) [inline travis.sh]: I think you are trying this: https://github.com/SciRuby/daru-view/pull/131/commits/d6de2718707bce40f4b2d9a21888c65c3e27f422

File: travis.sh [hunk 2]
```diff
@@ -45,12 +45,12 @@ then
       rbenv install --verbose $ruby_version
     )
 
-    gem pristine --all
-    gem update --no-document --system
-    gem update --no-document
+    yes | gem pristine --all
```
Comment (Shekharrajak) [inline travis.sh:48]: I didn't understand why it is required

File: .travis.yml [hunk 3]
```diff
@@ -80,50 +80,72 @@ matrix:
     # The latest stable and head versions of ruby built by clang on OSX
     - os: osx
       compiler: clang
-      rvm: 2.2
+      rvm: ruby-head-clang
```
Comment (Shekharrajak) [inline .travis.yml:83]: What was the reason behind these changes? Can you please try to write comments for your changes and what issues it is solving?

File: travis.sh (hunk 2 above)
Comment (Uditgulati) [reply travis.sh:48]: In some builds, while these `gem` commands are being executed, it asks "Overwrite the executable? [yN]" and it waits for `yes` or `no` input and after 10 mins travis times out. In order to automatically answer `yes` if such case occurs, `yes | ` is prepended.

File: travis.sh (hunk 1 above)
Comment (Uditgulati) [reply travis.sh]: Yeah, that'd be a better way of doing it. Thanks.

File: .travis.yml (hunk 3 above)
Comment (Uditgulati) [reply .travis.yml:83]: For osx builds, the clang version of corresponding ruby versions are used. Additional parameters are added so as to make sure that no error occurs during build. I'll add the comments accordingly.

**Comment by Uditgulati** [comment]
@Shekharrajak I've made some changes. Kindly have a look and let me know if anything else needs to be changed.

File: .travis.yml (hunk 3 above)
Comment (Shekharrajak) [reply .travis.yml]: Yeah!  That I understood. But I believe without also it is fine. Anyways it is not a problem.

File: travis.sh (hunk 2 above)
Comment (Shekharrajak) [reply travis.sh:48]: Yeah! That I understood. But I believe without also it is fine. Anyways it is not a problem.

Comment (Uditgulati) [reply travis.sh:48]: Do you need anything else to be modified?

**Comment by Uditgulati** [comment]
@Shekharrajak  Hi, 
I think we should merge this PR now as it does fix most of builds. It'll help prevent errors in user builds on systems with compatibly issues due to old versions.

**Comment by Shekharrajak** [comment]
Actually, we need to look into https://travis-ci.org/SciRuby/nmatrix/jobs/501367373 this travis failure as well. I think @prasunanand can review it better.

**Pull request closed without merging**